File: zero-server/src/main/kotlin/com/zeroday/service/ContractGenerator.kt
package com.zeroday.service

import com.zeroday.model.EnhancedContract
import com.zeroday.model.MissionModifiers
import com.zeroday.model.ModifierType
import com.zeroday.model.TaskDifficulty
import com.zeroday.model.TaskTargetType
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.UUID
import kotlin.random.Random

class ContractGenerator(
    private val playerService: PlayerService
) {
    data class ContractTemplate(
        val id: String,
        val titleTemplate: String,
        val descriptionTemplate: String,
        val targetType: TaskTargetType,
        val suggestedCommands: List<String>
    )

    private val mutex = Mutex()

    private val contractTemplates = listOf(
        ContractTemplate("data_breach", "Data Breach: {company}",
            "Infiltrate {target} and exfiltrate the {data_type} database. {difficulty_note}",
            TaskTargetType.DATA_THEFT, listOf("exploit", "ssh", "cat")),
        ContractTemplate("pen_test", "Penetration Test: {company}",
            "Perform a full security audit of {target}. Identify and exploit vulnerabilities. {difficulty_note}",
            TaskTargetType.PENETRATION_TEST, listOf("nmap", "exploit", "backdoor")),
        ContractTemplate("ddos_job", "Service Disruption: {company}",
            "Take {target} offline for {duration} seconds. Coordinate with available resources. {difficulty_note}",
            TaskTargetType.SERVICE_DISRUPTION, listOf("overload", "botnet")),
        ContractTemplate("recon", "Reconnaissance: {target_net}",
            "Map out the {target_net} subnet. Identify all hosts, open ports, and running services. {difficulty_note}",
            TaskTargetType.NETWORK_RECON, listOf("nmap", "ping", "traceroute")),
        ContractTemplate("crypto_job", "Cryptographic Challenge: {company}",
            "Intercept and decrypt {data_type} communications from {target}. Multiple encryption layers. {difficulty_note}",
            TaskTargetType.CRYPTOGRAPHY, listOf("decrypt", "sniff", "crack")),
        ContractTemplate("cleanup", "Log Cleanup: {company}",
            "Erase all traces of {prev_attacker} intrusion from {target}. Remove logs and backdoors. {difficulty_note}",
            TaskTargetType.FORENSICS, listOf("firewall", "trace", "worm")),
        ContractTemplate("botnet_recruit", "Botnet Recruitment Drive",
            "Infect {node_count} systems in {target_net} with backdoor payload. Build your botnet. {difficulty_note}",
            TaskTargetType.BOTNET_OPERATION, listOf("backdoor", "botnet", "worm")),
        ContractTemplate("social_engineer", "Social Engineering: {company}",
            "Set up a phishing campaign targeting {company} employees. Harvest {credential_count} credentials. {difficulty_note}",
            TaskTargetType.SOCIAL_ENGINEERING, listOf("proxy", "spoof", "encrypt")),
        ContractTemplate("bug_bounty", "Bug Bounty: {company}",
            "Find and document {vuln_count} vulnerabilities in {target}. Submit responsible disclosure. {difficulty_note}",
            TaskTargetType.BUG_BOUNTY, listOf("sqlmap", "exploit", "nmap"))
    )

    private val companyNames = listOf(
        "NexusCorp", "ApexGlobal", "ByteDynamics", "QuantumLeap", "FusionTech",
        "OmniData", "StratoSys", "PulseNet", "VertexLogic", "CipherStack",
        "MatrixHoldings", "EchoSystems", "DeltaForce Tech", "Sigma Solutions",
        "Omega Labs", "Phoenix Cyber", "Titan Industries", "Aurora Networks"
    )

    private val dataTypes = listOf("customer", "employee", "financial", "medical", "research", "classified")
    private val difficultyNotes: Map<TaskDifficulty, String> = mapOf(
        TaskDifficulty.TRIVIAL to "A cakewalk. Even a script kiddie could do this.",
        TaskDifficulty.EASY to "Entry-level contract. Basic security measures.",
        TaskDifficulty.MEDIUM to "Standard corporate security. Proceed with caution.",
        TaskDifficulty.HARD to "Enterprise-grade defenses. Bring your A-game.",
        TaskDifficulty.EXPERT to "Military-grade encryption and monitoring. Extreme caution advised.",
        TaskDifficulty.LEGENDARY to "Three-letter-agency level. You didn't see this. Good luck."
    )

    suspend fun generateContract(playerId: String): EnhancedContract? = mutex.withLock {
        val player = playerService.getPlayer(playerId) ?: return@withLock null

        val template = contractTemplates.random()
        val difficulty = rollDifficulty(player.level)
        val company = companyNames.random()
        val targetIp = generateTargetIp(player.level)

        val fillParams: Map<String, String> = mapOf(
            "company" to company,
            "target" to targetIp,
            "target_net" to "${targetIp.substringBeforeLast(".")}.0/24",
            "data_type" to dataTypes.random(),
            "duration" to "${(30 * difficulty.multiplier).toInt() + 10}",
            "node_count" to "${(5 * difficulty.multiplier).toInt() + 2}",
            "credential_count" to "${(5 * difficulty.multiplier).toInt()}",
            "vuln_count" to "${(2 * difficulty.multiplier).toInt().coerceAtLeast(2)}",
            "prev_attacker" to listOf("Anonymous", "GhostNet", "DarkOverlord", "LulzSec").random(),
            "difficulty_note" to (difficultyNotes[difficulty] ?: "")
        )

        val modifierCount = Random.nextInt(4)
        val modifiers = if (modifierCount > 0) {
            MissionModifiers.getRandomModifier(modifierCount)
        } else emptyList()

        var finalDifficulty = difficulty
        var rewardMultiplier = 1.0
        for (modifier in modifiers) {
            when (modifier.modifierType) {
                ModifierType.INCREASED_DIFFICULTY -> {
                    val currentIndex = finalDifficulty.ordinal
                    if (currentIndex < TaskDifficulty.entries.size - 1) {
                        finalDifficulty = TaskDifficulty.entries[currentIndex + 1]
                    }
                }
                ModifierType.DECREASED_DIFFICULTY -> {
                    val currentIndex = finalDifficulty.ordinal
                    if (currentIndex > 0) {
                        finalDifficulty = TaskDifficulty.entries[currentIndex - 1]
                    }
                }
                ModifierType.BOOSTED_REWARDS -> rewardMultiplier *= 1.5
                ModifierType.REDUCED_REWARDS -> rewardMultiplier *= 0.7
                ModifierType.SPEED_RUN -> rewardMultiplier *= 2.0
                ModifierType.TIME_LIMIT -> {}
                else -> {}
            }
        }

        val xpBase = (50 * difficulty.multiplier * player.level).toLong()
        val creditsBase = (100 * difficulty.multiplier * player.level).toLong()
        val finalXp = (xpBase * rewardMultiplier).toLong()
        val finalCredits = (creditsBase * rewardMultiplier).toLong()

        val title = fillTemplate(template.titleTemplate, fillParams)
        val description = fillTemplate(template.descriptionTemplate, fillParams)

        return@withLock EnhancedContract(
            id = UUID.randomUUID().toString(),
            title = title,
            description = description,
            targetType = template.targetType,
            targetIp = targetIp,
            requiredCommands = template.suggestedCommands,
            difficulty = finalDifficulty,
            suggestedLevel = player.level,
            baseXp = xpBase,
            baseCredits = creditsBase,
            modifiers = modifiers,
            finalRewardMultiplier = rewardMultiplier,
            finalXp = finalXp,
            finalCredits = finalCredits
        )
    }

    private fun rollDifficulty(playerLevel: Int): TaskDifficulty {
        val maxDiff = when {
            playerLevel < 3 -> 0
            playerLevel < 5 -> 1
            playerLevel < 7 -> 2
            playerLevel < 10 -> 3
            playerLevel < 14 -> 4
            else -> 5
        }
        val roll = Random.nextInt(maxDiff + 2)
        return TaskDifficulty.entries[roll.coerceAtMost(TaskDifficulty.entries.size - 1)]
    }

    private fun generateTargetIp(playerLevel: Int): String {
        val base = when {
            playerLevel < 3 -> "10.0.0"
            playerLevel < 5 -> "10.0.1"
            playerLevel < 7 -> "10.0.2"
            playerLevel < 10 -> "10.0.3"
            playerLevel < 14 -> "172.16.0"
            playerLevel < 18 -> "198.51.100"
            else -> "203.0.113"
        }
        return "$base.${Random.nextInt(2, 255)}"
    }

    private fun fillTemplate(template: String, params: Map<String, String>): String {
        var result = template
        params.forEach { (key, value) -> result = result.replace("{$key}", value) }
        return result
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Services/ContractGenerator.cs
namespace ZeroDayMMO.Server.Services;

public enum ModifierType
{
    TimeLimit,
    StealthRequired,
    NoDetection,
    SpeedRun,
    GhostMode,
    LimitedResources,
    BoostedRewards,
    ReducedRewards,
    IncreasedDifficulty,
    DecreasedDifficulty,
    SpecialObjective,
    TeamRequired,
    SoloOnly
}

public enum ModifierRarity
{
    Common,
    Uncommon,
    Rare,
    Epic,
    Legendary
}

public record MissionModifier(
    string Id,
    string Name,
    string Description,
    ModifierType ModifierType,
    double Value,
    ModifierRarity Rarity,
    long DurationMs = 0,
    bool Stacking = false
);

public record EnhancedContract(
    string Id,
    string Title,
    string Description,
    TaskTargetType TargetType,
    string TargetIp,
    List<string> RequiredCommands,
    TaskDifficulty Difficulty,
    int SuggestedLevel,
    long BaseXp,
    long BaseCredits,
    List<MissionModifier> Modifiers,
    double FinalRewardMultiplier,
    long FinalXp,
    long FinalCredits,
    long CreatedAt,
    long TimeLimitMs
);

public static class MissionModifiers
{
    public static readonly MissionModifier TimeLimited = new(
        "time_limited", "Time Limited", "Complete within reduced time",
        ModifierType.TimeLimit, 0.7, ModifierRarity.Uncommon);

    public static readonly MissionModifier StealthRequired = new(
        "stealth_required", "Stealth Required", "Avoid detection systems",
        ModifierType.StealthRequired, 1.0, ModifierRarity.Rare);

    public static readonly MissionModifier NoDetection = new(
        "no_detection", "Ghost Mode", "Leave no traces behind",
        ModifierType.NoDetection, 1.0, ModifierRarity.Epic);

    public static readonly MissionModifier SpeedRun = new(
        "speed_run", "Speed Run", "2x rewards for completing under half time",
        ModifierType.SpeedRun, 2.0, ModifierRarity.Rare);

    public static readonly MissionModifier LimitedResources = new(
        "limited_resources", "Limited Resources", "Reduced starting resources",
        ModifierType.LimitedResources, 0.5, ModifierRarity.Uncommon);

    public static readonly MissionModifier BoostedRewards = new(
        "boosted_rewards", "Boosted Rewards", "Increased payout for risk",
        ModifierType.BoostedRewards, 1.5, ModifierRarity.Uncommon);

    public static readonly MissionModifier ReducedRewards = new(
        "reduced_rewards", "Reduced Rewards", "Lower payout for ease",
        ModifierType.ReducedRewards, 0.7, ModifierRarity.Common);

    public static readonly MissionModifier IncreasedDifficulty = new(
        "increased_difficulty", "Increased Difficulty", "Enhanced security measures",
        ModifierType.IncreasedDifficulty, 1.0, ModifierRarity.Rare);

    public static readonly MissionModifier DecreasedDifficulty = new(
        "decreased_difficulty", "Decreased Difficulty", "Weakened defenses",
        ModifierType.DecreasedDifficulty, 1.0, ModifierRarity.Uncommon);

    public static readonly MissionModifier SpecialObjective = new(
        "special_objective", "Special Objective", "Additional objective required",
        ModifierType.SpecialObjective, 1.0, ModifierRarity.Rare);

    public static readonly MissionModifier TeamRequired = new(
        "team_required", "Team Required", "Must complete with 2+ players",
        ModifierType.TeamRequired, 1.0, ModifierRarity.Rare);

    public static readonly MissionModifier SoloOnly = new(
        "solo_only", "Solo Only", "Must be completed alone",
        ModifierType.SoloOnly, 1.0, ModifierRarity.Uncommon);

    public static List<MissionModifier> AllModifiers { get; } = new()
    {
        TimeLimited, StealthRequired, NoDetection, SpeedRun,
        LimitedResources, BoostedRewards, ReducedRewards,
        IncreasedDifficulty, DecreasedDifficulty, SpecialObjective,
        TeamRequired, SoloOnly
    };

    public static List<MissionModifier> GetRandomModifier(int count = 1)
    {
        var takeCount = Math.Min(count, AllModifiers.Count);
        return Enumerable.Range(0, takeCount)
            .Select(_ => AllModifiers[Random.Shared.Next(AllModifiers.Count)])
            .Distinct()
            .ToList();
    }
}

public class ContractGenerator
{
    public record ContractTemplate(
        string Id,
        string TitleTemplate,
        string DescriptionTemplate,
        TaskTargetType TargetType,
        List<string> SuggestedCommands
    );

    private readonly IPlayerService _playerService;
    private readonly SemaphoreSlim _mutex = new(1, 1);

    private readonly List<ContractTemplate> _contractTemplates = new()
    {
        new("data_breach", "Data Breach: {company}",
            "Infiltrate {target} and exfiltrate the {data_type} database. {difficulty_note}",
            TaskTargetType.DataTheft, new() { "exploit", "ssh", "cat" }),
        new("pen_test", "Penetration Test: {company}",
            "Perform a full security audit of {target}. Identify and exploit vulnerabilities. {difficulty_note}",
            TaskTargetType.PenetrationTest, new() { "nmap", "exploit", "backdoor" }),
        new("ddos_job", "Service Disruption: {company}",
            "Take {target} offline for {duration} seconds. Coordinate with available resources. {difficulty_note}",
            TaskTargetType.ServiceDisruption, new() { "overload", "botnet" }),
        new("recon", "Reconnaissance: {target_net}",
            "Map out the {target_net} subnet. Identify all hosts, open ports, and running services. {difficulty_note}",
            TaskTargetType.NetworkRecon, new() { "nmap", "ping", "traceroute" }),
        new("crypto_job", "Cryptographic Challenge: {company}",
            "Intercept and decrypt {data_type} communications from {target}. Multiple encryption layers. {difficulty_note}",
            TaskTargetType.Cryptography, new() { "decrypt", "sniff", "crack" }),
        new("cleanup", "Log Cleanup: {company}",
            "Erase all traces of {prev_attacker} intrusion from {target}. Remove logs and backdoors. {difficulty_note}",
            TaskTargetType.Forensics, new() { "firewall", "trace", "worm" }),
        new("botnet_recruit", "Botnet Recruitment Drive",
            "Infect {node_count} systems in {target_net} with backdoor payload. Build your botnet. {difficulty_note}",
            TaskTargetType.BotnetOperation, new() { "backdoor", "botnet", "worm" }),
        new("social_engineer", "Social Engineering: {company}",
            "Set up a phishing campaign targeting {company} employees. Harvest {credential_count} credentials. {difficulty_note}",
            TaskTargetType.SocialEngineering, new() { "proxy", "spoof", "encrypt" }),
        new("bug_bounty", "Bug Bounty: {company}",
            "Find and document {vuln_count} vulnerabilities in {target}. Submit responsible disclosure. {difficulty_note}",
            TaskTargetType.BugBounty, new() { "sqlmap", "exploit", "nmap" })
    };

    private readonly List<string> _companyNames = new()
    {
        "NexusCorp", "ApexGlobal", "ByteDynamics", "QuantumLeap", "FusionTech",
        "OmniData", "StratoSys", "PulseNet", "VertexLogic", "CipherStack",
        "MatrixHoldings", "EchoSystems", "DeltaForce Tech", "Sigma Solutions",
        "Omega Labs", "Phoenix Cyber", "Titan Industries", "Aurora Networks"
    };

    private readonly List<string> _dataTypes = new()
    {
        "customer", "employee", "financial", "medical", "research", "classified"
    };

    private readonly Dictionary<TaskDifficulty, string> _difficultyNotes = new()
    {
        [TaskDifficulty.Trivial] = "A cakewalk. Even a script kiddie could do this.",
        [TaskDifficulty.Easy] = "Entry-level contract. Basic security measures.",
        [TaskDifficulty.Medium] = "Standard corporate security. Proceed with caution.",
        [TaskDifficulty.Hard] = "Enterprise-grade defenses. Bring your A-game.",
        [TaskDifficulty.Expert] = "Military-grade encryption and monitoring. Extreme caution advised.",
        [TaskDifficulty.Legendary] = "Three-letter-agency level. You didn't see this. Good luck."
    };

    private readonly List<string> _prevAttackers = new()
    {
        "Anonymous", "GhostNet", "DarkOverlord", "LulzSec"
    };

    public ContractGenerator(IPlayerService playerService)
    {
        _playerService = playerService;
    }

    public async Task<EnhancedContract?> GenerateContract(string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            var player = _playerService.GetPlayer(playerId);
            if (player == null) return null;

            var template = _contractTemplates[Random.Shared.Next(_contractTemplates.Count)];
            var difficulty = RollDifficulty(player.Level);
            var company = _companyNames[Random.Shared.Next(_companyNames.Count)];
            var targetIp = GenerateTargetIp(player.Level);

            var dataType = _dataTypes[Random.Shared.Next(_dataTypes.Count)];
            var prevAttacker = _prevAttackers[Random.Shared.Next(_prevAttackers.Count)];

            var fillParams = new Dictionary<string, string>
            {
                ["company"] = company,
                ["target"] = targetIp,
                ["target_net"] = targetIp[..targetIp.LastIndexOf('.')] + ".0/24",
                ["data_type"] = dataType,
                ["duration"] = $"{(int)(30 * difficulty.Multiplier()) + 10}",
                ["node_count"] = $"{(int)(5 * difficulty.Multiplier()) + 2}",
                ["credential_count"] = $"{(int)(5 * difficulty.Multiplier())}",
                ["vuln_count"] = $"{Math.Max(2, (int)(2 * difficulty.Multiplier()))}",
                ["prev_attacker"] = prevAttacker,
                ["difficulty_note"] = _difficultyNotes.GetValueOrDefault(difficulty, "")
            };

            var modifierCount = Random.Shared.Next(4);
            var modifiers = modifierCount > 0
                ? MissionModifiers.GetRandomModifier(modifierCount)
                : new List<MissionModifier>();

            var finalDifficulty = difficulty;
            var rewardMultiplier = 1.0;

            foreach (var modifier in modifiers)
            {
                switch (modifier.ModifierType)
                {
                    case ModifierType.IncreasedDifficulty:
                        var currentIdx = (int)finalDifficulty;
                        if (currentIdx < Enum.GetValues<TaskDifficulty>().Length - 1)
                            finalDifficulty = (TaskDifficulty)(currentIdx + 1);
                        break;
                    case ModifierType.DecreasedDifficulty:
                        var curIdx = (int)finalDifficulty;
                        if (curIdx > 0)
                            finalDifficulty = (TaskDifficulty)(curIdx - 1);
                        break;
                    case ModifierType.BoostedRewards:
                        rewardMultiplier *= 1.5;
                        break;
                    case ModifierType.ReducedRewards:
                        rewardMultiplier *= 0.7;
                        break;
                    case ModifierType.SpeedRun:
                        rewardMultiplier *= 2.0;
                        break;
                }
            }

            var xpBase = (long)(50 * difficulty.Multiplier() * player.Level);
            var creditsBase = (long)(100 * difficulty.Multiplier() * player.Level);
            var finalXp = (long)(xpBase * rewardMultiplier);
            var finalCredits = (long)(creditsBase * rewardMultiplier);

            var title = FillTemplate(template.TitleTemplate, fillParams);
            var description = FillTemplate(template.DescriptionTemplate, fillParams);

            return new EnhancedContract(
                Id: Guid.NewGuid().ToString(),
                Title: title,
                Description: description,
                TargetType: template.TargetType,
                TargetIp: targetIp,
                RequiredCommands: new(template.SuggestedCommands),
                Difficulty: finalDifficulty,
                SuggestedLevel: player.Level,
                BaseXp: xpBase,
                BaseCredits: creditsBase,
                Modifiers: new(modifiers),
                FinalRewardMultiplier: rewardMultiplier,
                FinalXp: finalXp,
                FinalCredits: finalCredits,
                CreatedAt: DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                TimeLimitMs: 600_000
            );
        }
        finally
        {
            _mutex.Release();
        }
    }

    private TaskDifficulty RollDifficulty(int playerLevel)
    {
        var maxIdx = playerLevel switch
        {
            < 3 => 0,
            < 5 => 1,
            < 7 => 2,
            < 10 => 3,
            < 14 => 4,
            _ => 5
        };
        var roll = Random.Shared.Next(maxIdx + 2);
        var diffs = Enum.GetValues<TaskDifficulty>();
        return diffs[Math.Min(roll, diffs.Length - 1)];
    }

    private static string GenerateTargetIp(int playerLevel)
    {
        var baseIp = playerLevel switch
        {
            < 3 => "10.0.0",
            < 5 => "10.0.1",
            < 7 => "10.0.2",
            < 10 => "10.0.3",
            < 14 => "172.16.0",
            < 18 => "198.51.100",
            _ => "203.0.113"
        };
        return $"{baseIp}.{Random.Shared.Next(2, 255)}";
    }

    private static string FillTemplate(string template, Dictionary<string, string> parameters)
    {
        var result = template;
        foreach (var (key, value) in parameters)
        {
            result = result.Replace($"{{{key}}}", value);
        }
        return result;
    }
}
